Add spi_device (#3156)
OttoWinter/esphomedocs: components/spi.rst

- **clk_pin** (**Required**, :ref:`Pin Schema <config-pin_schema>`): The pin used for the clock line of the SPI bus.
- **mosi_pin** (*Optional*, :ref:`Pin Schema <config-pin_schema>`): The pin used for the MOSI line of the SPI bus.
- **miso_pin** (*Optional*, :ref:`Pin Schema <config-pin_schema>`): The pin used for the MISO line of the SPI bus.
- **id** (*Optional*, :ref:`config-id`): Manually specify the ID for this SPI hub if you need multiple SPI hubs.
- **interface** (*Optional*): Controls which hardware or software SPI interface should be used.
  Value may be one of ``any`` (default), ``software``, ``hardware``, ``spi``, ``spi2`` or ``spi3``, depending on
  the particular chip. See discussion below.
- **force_sw** (*Optional*, **Deprecated**, boolean): Whether software implementation should be used even if a hardware
  interface is available. Default is ``false``.

Interface selection:
--------------------

ESP32 and ESP8266 chips have several hardware SPI controller interfaces - usually the first one or two are reserved for use to access

> the flash and PSRAM memories, leaving one or two user-accessible SPI controllers. An SPI hub configured in

Reviewer: ```suggestion
the flash and PSRAM memories, leaving one or two user-accessible SPI controllers. A SPI hub configured in
```

Author: I pronounce "SPI" as "ess pee eye" which would require "An" before it (starts with a vowel sound.) I'll reword it to avoid the question :-)

WRT to the other comments, I'll clean them up - I'm going to use this nomenclature (avoiding "hub" which is more an internal term from what I've seen.)

* Component - the singleton SPI component, with one or more controllers;
* Controller - a hardware or software controller associated with a bus (set of data/control lines.)
* Device - a device attached to a controller via its bus.
* Interface - the controller implementation (hw or sw).

Reviewer: Sounds good! Thanks again. 🍻

ESPHome can be assigned to one of these interfaces with the ``interface:`` configuration option.

By default (``interface: any``) the first available hardware SPI interface will be assigned, a second if available then

> any further spi interfaces configured will use software mode. You can choose a specific interface with ``spi`` (meaning

Reviewer: It seems weird to say "SPI interface"... "Serial Peripheral Interface interface"...haha. Reworded this a little.
```suggestion
By default (``interface: any``) the first available SPI hardware interface will be assigned, then a second if available;
any additional interfaces configured will use software mode. You can choose a specific interface with ``spi`` (meaning
```

the first or only available interface) or one of ``spi2`` and ``spi3`` for ESP32 chips with two available SPI
controllers. Note that SPI0 and SPI1 are typically not available, being reserved for accessing flash and PSRAM.

If the ``software`` option is chosen, or you configure more SPI hubs than there are available hardware controllers,
the remaining hubs will use a software implementation, which is unable to achieve data rates above a few hundred

> kHz.

Reviewer: Calling this out...
```suggestion
kHz. This implementation may be reasonable for sensors but is not recommended for use with displays.
```

While the ESP32 supports the reassignment of the default SPI pins to most other GPIO pins, using the dedicated SPI pins
can improve performance and stability for certain ESP/device combinations.

> ESP8266 has a more limited selection of pins that can be used, again
> check the datasheet for more information.

Reviewer: ```suggestion
ESP8266 has a more limited selection of pins that can be used; check the datasheet for more information.
```
